Fix compilation errors and optimize stuff

stilic opened the pull request:
(My first actual good pull request to this repo...)

- Fix compilation errors and a few warnings
- Optimize usage of GPU (store images in the VRAM instead of the RAM)
- Optimize audio playing (play music while loading it at the same time)
- Change the Lua wrapper to [`hxlua`](https://github.com/MAJigsaw77/hxlua) for allowing videos to play on Linux
- Switch to the `main` Git branch of HxCodec (for the previous change to work)
- Fix the parsing of sustain notes from charts
- Fix sutain notes clipping (thanks to @NeeEoo for the fix)

stilic committed:
Fix compilation

stilic committed:
Optimize GPU and audio stuff

stilic committed:
Fix the defaultCameras warning

ThePlank commented:
not on expermimental branch <img src="https://cdn.discordapp.com/emojis/897578872344813610.webp?size=128&quality=lossless" height="15em"/>

Stilic commented:
> not on expermimental branch <img alt="" height="15em" src="https://camo.githubusercontent.com/584e358d017363946cd794c36111e12d15d46529646f1d508095121deca95b89/68747470733a2f2f63646e2e646973636f72646170702e636f6d2f656d6f6a69732f3839373537383837323334343831333631302e776562703f73697a653d313238267175616c6974793d6c6f73736c657373">

Trust me, most of the best pull requests (even the one that MasterEric, a dev of the official game made) are based on the `main` branch.

Stilic commented:
And if it's really needed, I can rebase my fork to that branch.

stilic committed:
Update HxCodec version

Stilic commented:
> Been trying to add the cache to gpu myself for a while, all I can say is that I went the completely wrong way about it. This is excellent, decreased RAM usage A LOT.
> 
> However, probably wouldn't have it on by default, since from what I've seen the majority of FNF players run with weak integrated laptop GPUs which usually allocates only 1GB (or less!) of the ram as VRAM. A lot of the mods that are done using LUA easily breaches 1GB, this would result in severe freezing when the allocated memory gets full since the driver needs to clean and refill the cache.

What about doing a check of the VRAM capacity? (if it is below or equal to 1 GB, disable the feature)

lemz1 commented:
idk if this would affect performance in a bad way, but maybe check on each Paths.image (or whatever) call if the VRAM usage exceeds a chosen percentage (for example 80%).
if it does exceed the percentage, then use RAM instead of VRAM
this would i guess make it so that both VRAM and RAM are used.
but idk much about memory usage, so this might be a bad idea.

Stilic commented:
> idk if this would affect performance in a bad way, but maybe check on each Paths.image (or whatever) call if the VRAM usage exceeds a chosen percentage (for example 80%). if it does exceed the percentage, then use RAM instead of VRAM this would i guess make it so that both VRAM and RAM are used. but idk much about memory usage, so this might be a bad idea.

This seems to be impossible to do.
I didn't found a way to get the total amount of VRAM with Haxe.

stilic committed:
Add proper video support on Linux
Also change the whole Lua library to something that could works with hxCodec

stilic committed:
Merge pull request #1 from Stilic/video-linux

Add proper video support on Linux

stilic committed:
Disable GPU caching by default
Suggested by https://github.com/ShadowMario/FNF-PsychEngine/pull/12561#pullrequestreview-1473386486

stilic committed:
Fix sustain notes parsing

Commit:
Update EditorPlayState.hx

fix "Section" labeled as "Beat"

stilic committed:
Merge pull request #2 from BeastlyGabi/patch-6

Update EditorPlayState.hx

Luisinhi010 commented:
> > idk if this would affect performance in a bad way, but maybe check on each Paths.image (or whatever) call if the VRAM usage exceeds a chosen percentage (for example 80%). if it does exceed the percentage, then use RAM instead of VRAM this would i guess make it so that both VRAM and RAM are used. but idk much about memory usage, so this might be a bad idea.
> 
> This seems to be impossible to do. I didn't found a way to get the total amount of VRAM with Haxe. I guess I can just disable GPU caching by default.

try this: FlxG.stage.context3D.totalGPUMemory;

lemz1 commented:
> try this: FlxG.stage.context3D.totalGPUMemory;

this only gives you the current allocated gpu memory
not the total vram size

Luisinhi010 commented:
oh yeah

> > try this: FlxG.stage.context3D.totalGPUMemory;
> 
> this only gives you the current allocated gpu memory not the total vram size

MemeHovy commented:
> oh yeah
> 
> > > try this: FlxG.stage.context3D.totalGPUMemory;
> > 
> > 
> > this only gives you the current allocated gpu memory not the total vram size

hi Luis

Luisinhi010 commented:
> > oh yeah
> > 
> > > > try this: FlxG.stage.context3D.totalGPUMemory;
> > > 
> > > 
> > > this only gives you the current allocated gpu memory not the total vram size
> 
> hi Luis 

Hello Hoovy

Rom4chek commented:
Oh my, nice one

KookerFoxYT commented:
> Reference in ne

it is, because v0.7 is already coming out soon sooo

ACoolioDude reviewed (approved):
Oh wow

Sword352 reviewed (approved):
Would be better if it was on experimental as 0.7 will be pushed to main but cool stuff

stilic committed:
Fix sustain clipping

Stefan2008Git commented:
![slika](https://github.com/ShadowMario/FNF-PsychEngine/assets/126715034/93ca7509-7ef2-4fdf-81f9-3203141b918d)
Buddy, your build was failed. Use hxCodec 2.5.1 LMAO or use this line:

`#if VIDEOS_ALLOWED`
`#if (hxCodec >= "3.0.0")`
`import hxcodec.flixel.FlxVideo as MP4Handler;`
`#elseif (hxCodec >= "2.6.1")`
`import hxcodec.VideoHandler as MP4Handler;`
`#elseif (hxCodec == "2.6.0")`
`import VideoHandler as MP4Handler;`
`#else`
`import vlc.MP4Handler;`
`#end`
`#end`

Stilic commented:
> ![slika](https://user-images.githubusercontent.com/126715034/250993955-93ca7509-7ef2-4fdf-81f9-3203141b918d.png) Buddy, your build was failed. Use hxCodec 2.5.1 LMAO or use this line:
> 
> `#if VIDEOS_ALLOWED` `#if (hxCodec >= "3.0.0")` `import hxcodec.flixel.FlxVideo as MP4Handler;` `#elseif (hxCodec >= "2.6.1")` `import hxcodec.VideoHandler as MP4Handler;` `#elseif (hxCodec == "2.6.0")` `import VideoHandler as MP4Handler;` `#else` `import vlc.MP4Handler;` `#end` `#end`

I'll fix that in a moment.

stilic committed:
Fix HxCodec

Stefan2008Git commented:
Now Mac build has have error LMAO.

Stefan2008Git commented:
> Now Mac build has have error LMAO.

![Screenshot_20230705-111013~2](https://github.com/ShadowMario/FNF-PsychEngine/assets/126715034/e1e196f7-6521-46cb-958d-d9225ab62c99)

Oh lol. Look at that

stilic committed:
Fix HxCodec again

BeastlyGhost commented:
this should be rebased to experimental branch since its currently in pre-release phase

Stilic commented:
> 

I'm going to do that later since i'm busy for now.
